frontend/src/components/MinioArtifactLink.tsx

```diff
@@ -1,30 +1,36 @@
 import * as React from 'react';
-import { StoragePath, StorageService } from '../lib/WorkflowParser';
+import { StorageService } from '../lib/WorkflowParser';
 import { S3Artifact } from '../../third_party/argo-ui/argo_template';
-
-const artifactApiUri = ({ source, bucket, key }: StoragePath) =>
-  'artifacts/get' + `?source=${source}&bucket=${bucket}&key=${encodeURIComponent(key)}`;
+import { generateArtifactUrl } from '../lib/Utils';
+import { Apis } from '../lib/Apis';
 
 /**
- * A component that renders an artifact link.
+ * A component that renders a preview to an artifact with a link to the full content.
  */
 const MinioArtifactLink: React.FC<S3Artifact> = s3artifact => {
-  if (!s3artifact || !s3artifact.key || !s3artifact.bucket) {
+  const { key, bucket, endpoint } = s3artifact || {};
+  if (!key || !bucket) {
     return null;
   }
 
-  const { key, bucket, endpoint } = s3artifact;
+  const peek = 100;
+  const encodedKey = encodeURIComponent(key);
   const source = endpoint === 's3.amazonaws.com' ? StorageService.S3 : StorageService.MINIO;
-  const linkText = `${source.toString()}://${bucket}/${key}`;
+  const linkText = `${source.toString()}://${bucket}/${encodedKey}`;
+  const url = generateArtifactUrl(source, bucket, encodedKey);
+  const [content, setContent] = React.useState(linkText);
+
+  React.useEffect(() => {
+    Apis.readFile({ source, bucket, key }, peek).then(
+      data => setContent(data.length > peek ? `${data} ...` : data),
+      error => setContent(error),
+    );
+  }, [source, bucket, key]);
+
   // Opens in new window safely
   return (
-    <a
-      href={artifactApiUri({ key, bucket, source })}
-      target={'_blank'}
-      rel={'noreferrer noopener'}
-      title={linkText}
-    >
-      {linkText}
+    <a href={url} target={'_blank'} rel={'noreferrer noopener'} title={linkText}>
+      {content}
```

Comment:
Some opinions about this UX:
1. When UI server doesn't have permission to fetch content (that's a common deployment problem), this should better at least show user the link.
2. The link is already known, while content takes some time to fetch. IMO, we should better show the link before fetching results.

With my two opinions, I'm thinking about the following UX:
```
[link]
```
when loaded, it becomes
```
[link]
content
```

What do you think about this?

Reply:
Yup this is a good idea. Originally I was thinking of extending it abit more.

i.e. if the payload is sufficiently small, to actually render the json/csv/etc.

But let's keep it simple for now.

Reply:
SG, we can build more features later